### PR TITLE
[Backend][SAML] SAML Support 

### PR DESCRIPTION
**Pull request by mohityadav766**
### Describe your changes :

Saml support #2664

#
### Type of change :

- [x] Improvement
- [x] New feature

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#

### CODE REVIEW

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/CatalogApplication.java [hunk 1]
```diff
@@ -150,10 +155,26 @@ public void run(CatalogApplicationConfig catalogConfig, Environment environment)
...
+    registerSamlHandlers(catalogConfig, environment);
+
     FilterRegistration.Dynamic micrometerFilter =
         environment.servlets().addFilter("MicrometerHttpFilter", new MicrometerHttpFilter());
     micrometerFilter.addMappingForUrlPatterns(EnumSet.allOf(DispatcherType.class), true, "/*");
-    intializeWebsockets(catalogConfig, environment);
+  }
+
+  private void registerSamlHandlers(CatalogApplicationConfig catalogConfig, Environment environment) {
+    SamlSettingsHolder.getInstance().initDefaultSettings(catalogConfig);
```
Comment (harshach) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/CatalogApplication.java:168]: we should only add these settings if the catalogConfig has same settings

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/resources/teams/UserResource.java [hunk 2]
```diff
@@ -371,6 +371,24 @@ public Response createUser(
...
+      description = "Logout a User",
+      responses = {
+        @ApiResponse(
+            responseCode = "200",
+            description = "The user ",
+            content = @Content(mediaType = "application/json", schema = @Schema(implementation = User.class))),
+        @ApiResponse(responseCode = "400", description = "Bad request")
+      })
+  public Response logoutUser(@Context UriInfo uriInfo, @Context SecurityContext securityContext) throws IOException {
+    return Response.status(200).build();
```
Comment (harshach) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/resources/teams/UserResource.java:389]: we are not doing anything here?

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/JwtFilter.java [hunk 3]
```diff
@@ -103,13 +110,27 @@ public void filter(ContainerRequestContext requestContext) {
     String tokenFromHeader = extractToken(headers);
     LOG.debug("Token from header:{}", tokenFromHeader);
 
+    // the case where OMD generated the Token for the Client
+    if (providerType.equals("saml") || providerType.equals("basic")) {
```
Comment (harshach) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/JwtFilter.java:114]: please add final strings for all the constants

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/JwtFilter.java [hunk 4]
```diff
@@ -103,13 +110,27 @@ public void filter(ContainerRequestContext requestContext) {
...
+    }
+
     DecodedJWT jwt = validateAndReturnDecodedJwtToken(tokenFromHeader);
 
     Map<String, Claim> claims = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
     claims.putAll(jwt.getClaims());
 
     String userName = validateAndReturnUsername(claims);
 
+    if (uriInfo.getPath().contains("logout") && (providerType.equals("saml") || providerType.equals("basic"))) {
```
Comment (harshach) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/JwtFilter.java:125]: same here

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/jwt/JWTTokenGenerator.java [hunk 5]
```diff
@@ -91,9 +94,35 @@ public JWTAuthMechanism generateJWTToken(User user, JWTTokenExpiry expiry) {
...
+  public JWTAuthMechanism generateJWTToken(String userName, String email, JWTTokenExpiry expiry, boolean isBot) {
+    try {
+      JWTAuthMechanism jwtAuthMechanism = new JWTAuthMechanism().withJWTTokenExpiry(expiry);
+      Algorithm algorithm = Algorithm.RSA256(null, privateKey);
+      Date expires = getExpiryDate(expiry);
+      String token =
+          JWT.create()
+              .withIssuer(issuer)
+              .withKeyId(kid)
+              .withClaim("sub", userName)
```
Comment (harshach) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/jwt/JWTTokenGenerator.java:106]: same

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/resources/teams/UserResource.java (hunk 2 above)
Comment (harshach) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/resources/teams/UserResource.java:389]: why not invalidate the token here?

**Review by harshach — CHANGES_REQUESTED** [review]
@mohitdeuex left few comments take a look

**Comment by harshach** [comment]
@mohityadav766 guessing this is not needed. Closing.

**Comment by dpadmor** [comment]
@harshach why is not needed ??

**Pull request closed without merging**